components/speedreader/rust/lib/src/speedreader_readability.rs

```diff
@@ -0,0 +1,82 @@
+use html5ever::driver::{ParseOpts, Parser};
+use html5ever::tendril::{StrTendril, TendrilSink};
+use kuchiki::Sink;
+use lol_html::OutputSink;
+use std::collections::HashMap;
+use url::Url;
+
+use super::speedreader::*;
+
+use readability::{extractor, statistics};
+
+const MOZ_SCORE_THRESHOLD: f64 = 30.0;
+
+pub struct SpeedReaderReadability<O>
+where
+    O: OutputSink,
+{
+    parser: Parser<Sink>,
+    url: Url,
+    output_sink: O,
+}
+
+impl<O: OutputSink> SpeedReaderProcessor for SpeedReaderReadability<O> {
+    fn write(&mut self, input: &[u8]) -> Result<(), SpeedReaderError> {
+        match StrTendril::try_from_byte_slice(input) {
+            Ok(tendril) => {
+                self.parser.process(tendril);
+                Ok(())
+            }
+            Err(_) => Err(SpeedReaderError::DocumentParseError(
+                "Could not write tendril".to_owned(),
+            )),
+        }
+    }
+
+    fn end(&mut self) -> Result<(), SpeedReaderError> {
+        let dummy = html5ever::parse_document(Sink::default(), ParseOpts::default());
+        // NOTE: parser.finish() takes ownership, so we need this workaround
+        // since Parser<Sink> has no default.
+        let parser = std::mem::replace(&mut self.parser, dummy);
+        let mut dom: Sink = parser.finish();
```

Comment:
Yeah, i wish we could take `Self` but the FFI code has some assumptions about this. C++ passes it in as `void*`, and it gets cast to `Box<Box<dyn SpeedReaderProcessor>>`. I can't move out of the box. It gets freed from a different method. I had the `Option` version first, but I liked this more because at least it can be re-used.

Reply:
:grimacing: well that's fun!

> I had the Option version first, but I liked this more because at least it can be re-used.

Wouldn't this still not be reusable though? You'd probably want to return `Err(())` if it's already been `end`ed since the results won't make sense on subsequent calls.

Reply:
fixed